KAFKA-5597 Alternate way to do Metrics docs generation

wushujames opened the pull request:
I was about to start on the next round of autogeneration of metrics docs, but I wanted to @guozhangwang 's opinion on this first. This is a possible alternate way to do autogeneration of metrics docs, that possibly looks a little nicer for the developer. Posting this to get some feedback on if the original way looks better, or if this new way looks better.

Instead of having the metrics registry and the org.apache.kafka.common.metrics.Metrics object be separate things, have the metrics registry hold a copy of the Metrics object. That way, all the metricInstance stuff is hidden, and we don't have to make sure that the metrics registry and the Metrics object are configured identicailly (with the same tags).

I personally think this looks a little better.

wushujames committed:
Instead of having the metrics registry and the org.apache.kafka.common.metrics.Metrics object be separate things, have the metrics registry hold a copy of the Metrics object. That way, all the metricInstance stuff is hidden, and we don't have to make sure they are configured identicailly (with the same tags).

wushujames commented:
@guozhangwang Can I get your thoughts on this? I'd like to get the next round of metrics docs into 1.0.0, and I'd like some feedback on whether I should change my technique or not.

guozhangwang commented:
I like the idea of consolidating the logic in a single place / class. cc @ijuma to take another look.

ijuma commented:
Also cc @rajinisivaram who's done a bunch of metrics work recently

wushujames committed:
Merge remote-tracking branch 'origin/trunk' into producer_sender_metrics_docs_different

Conflicts:
	clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java
	clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java
	clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java

wushujames committed:
Fix merge errors.
Change public metrics to private, since they all have getters now.
Simplify the Sender.throttleTimeSensor() static method to just use the metrics registry.

wushujames committed:
Missed this change, when I switched metrics accces to private.

wushujames commented:
I merged in the latest trunk.

I also changed the MetricNameTemplate references to private, which means metrics can only be obtained  using accessors. The effect of this is that users of the MetricsRegistry objects can *only* obtained specific named metrics, which themselves were derived from the templates. Which kind of achieves my main goal, which is that all metrics have to be predefined and therefore will have their documentation autogenerated.

This works so well, in fact, that in Sender.java, you can see that it doesn't even have to import org.apache.kafka.common.metrics.Metrics anymore.

wushujames committed:
Merge remote-tracking branch 'origin/trunk' into producer_sender_metrics_docs_different

wushujames committed:
Update test case to accomodate new registry design.

wushujames committed:
Automatically put all the MetricNameTemplates into a list, which is used for documentation.

wushujames commented:
@guozhangwang: I think this PR is done, and I'd like to submit it for merge in order to get it into 1.0.0. Do I submit a new PR without "[WIP]"? Or do I just edit the title of the PR?

@ijuma @rajinisivaram: In this PR, I also improved the MetricsRegistry class so that you don't need to keep the getAllTemplates() array in sync with the class variables anymore.

Btw, after I get this PR in, I have an equivalent PR that refactors the ConsumerMetrics to follow the same pattern. So, heads up for that. Should I file a JIRA for that?

Thanks.

ijuma commented:
You can edit the title of this PR. Sure, feel free to file a JIRA for the `ConsumerMetrics` side.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java:
```diff
@@ -16,134 +16,253 @@
+    
+    private Metrics metrics;
+    private Set<String> tags;
+    private HashSet<String> topicTags;
```
Seems like most (all?) fields should be `final`.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java:
```diff
@@ -16,134 +16,253 @@
+
+    }
+
+    public MetricName getBatchSizeAvg() {
```
We generally avoid `getters` in Kafka.

wushujames replied:
In this case, the getters aren't standard "getters" which just return the internal variable. These getters are actually instantiating the MetricName object.

For many MetricNames, this isn't *entire* necessary since they are the same each time. But, for MetricNames that have tags (like per-topic MetricNames), it's necessary to instantiate them at get-time, because the tag may vary per metric.

wushujames replied:
@ijuma: Thanks for the feedback. Since you made me realize the getters aren't necessary all of the time, and you also pointed out that almost all fields should be final, I came up with a way to get rid of the getters, which also gets rid of a lot of the boilerplate code. I'll upload that shortly, so you can see how it looks.

wushujames committed:
Get rid of the getters, for all the metrics without additional tags. Precompute the MetricNames in advance. Expose those MetricNames as public variables to all the people who need it.

wushujames committed:
Inline access to the metrics, where possible and readable.

wushujames committed:
Make instance variables final.

wushujames committed:
Change another variable to final

wushujames committed:
Merge remote-tracking branch 'origin/trunk' into producer_sender_metrics_docs_different_2

wushujames commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -723,73 +718,60 @@ public static Sensor throttleTimeSensor(Metrics metrics, SenderMetricsRegistry m
-            this.batchSizeSensor.add(m, new Avg());
-            m = metrics.metricInstance(metricsRegistry.batchSizeMax);
-            this.batchSizeSensor.add(m, new Max());
+            this.batchSizeSensor.add(metrics.batchSizeAvg, new Avg());
```
Look at how clean and concise it is now. Direct variable access, and no need to call any methods, no need to use the Metrics object, and nothing is exposed about the metricsInstance or MetricNameTemplates.

wushujames commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -802,31 +784,31 @@ private void maybeRegisterTopicMetrics(String topic) {
                 topicRecordCount = this.metrics.sensor(topicRecordsCountName);
-                MetricName rateMetricName = this.metrics.metricInstance(metricsRegistry.topicRecordSendRate, metricTags);
-                MetricName totalMetricName = this.metrics.metricInstance(metricsRegistry.topicRecordSendTotal, metricTags);
+                MetricName rateMetricName = this.metrics.topicRecordSendRate(metricTags);
```
The ones above are accessing a variable, whereas this is calling a method. This is needed because I need to pass in the tags.

wushujames commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java:
```diff
@@ -16,134 +16,170 @@
+
+    /** topic level metrics **/
+    public MetricName topicRecordSendRate(Map<String, String> metricTags) {
+        return this.metrics.metricInstance(this.topicRecordSendRate, metricTags);
```
Notice that this method is named the same as the instance variable that it is using. Is that okay?

wushujames commented:
Test failure is not due to this PR. The failure is on trunk.

```
org.apache.kafka.common.protocol.ErrorsTest > testUniqueErrorCodes FAILED
    java.lang.AssertionError: Error codes must be unique expected:<61> but was:<62>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.apache.kafka.common.protocol.ErrorsTest.testUniqueErrorCodes(ErrorsTest.java:38)
```

It was introduced in https://github.com/apache/kafka/commit/5f6393f9b17cce17ded7a00e439599dfa77deb2d#diff-b119227df7efa3ffeb7fe69e49ff1afeR541

wushujames commented:
Hopefully trunk will be fixed soon. When it is, I will update this PR.

Can I get this merged to trunk, as well as to 1.0.0? I want to try to get https://issues.apache.org/jira/browse/KAFKA-5951 in, and it will layer on top of this PR.

wushujames committed:
Merge remote-tracking branch 'origin/trunk' into producer_sender_metrics_docs_different

guozhangwang commented:
As @ijuma replied on the mailing list, I think it should be fine to get in 1.0.0 if it is in mergable state in a week.

wushujames committed:
Merge remote-tracking branch 'origin/trunk' into producer_sender_metrics_docs_different

wushujames commented:
@guozhangwang @ijuma: This is ready for review.

wushujames commented:
@guozhangwang @ijuma Can you review? I'd like to get this and https://issues.apache.org/jira/browse/KAFKA-5951 into 1.0.0. My PR for https://issues.apache.org/jira/browse/KAFKA-5951 is ready but it layers on top of the changes in this PR, and so I can't submit it yet. I'd like to make sure that I have enough time to get it reviewed so it can merge in to 1.0.0. Thanks!

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerMetrics.java:
```diff
@@ -17,19 +17,22 @@
 public class ProducerMetrics {
 
     public SenderMetricsRegistry senderMetrics;
+    private Metrics metrics;
```
Both should be `final`.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -729,73 +724,60 @@ public static Sensor throttleTimeSensor(Metrics metrics, SenderMetricsRegistry m
-            MetricName totalMetricName = metrics.metricInstance(metricsRegistry.recordSendTotal);
+            MetricName rateMetricName = metrics.recordSendRate;
+            MetricName totalMetricName = metrics.recordSendTotal;
             this.recordsPerRequestSensor.add(new Meter(rateMetricName, totalMetricName));
```
Maybe inline this and other similar cases?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java:
```diff
@@ -16,134 +16,170 @@
-
-        this.produceThrottleTimeAvg = new MetricNameTemplate("produce-throttle-time-avg", METRIC_GROUP_NAME, "The average time in ms a request was throttled by a broker", tags);
-        this.produceThrottleTimeMax = new MetricNameTemplate("produce-throttle-time-max", METRIC_GROUP_NAME, "The maximum time in ms a request was throttled by a broker", tags);
+        this.batchSizeAvg = createMetricName("batch-size-avg", METRIC_GROUP_NAME, "The average number of bytes sent per partition per-request.", tags);
```
We can probably have one helper method for the `METRIC_GROUP_NAME` case and another for the `TOPIC_GROUP_NAME` case. Since these methods are called so many times, it's worth a little convenience. Also, the lines are too long at the moment. They should fit in the GitHub review window.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java:
```diff
@@ -16,134 +16,170 @@
+        return this.metrics.metricInstance(this.topicRecordErrorTotal, metricTags);
     }
 
     public List<MetricNameTemplate> getAllTemplates() {
```
Rename to `allTemplates()`?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java:
```diff
@@ -16,134 +16,170 @@
+    }
+
+    /** topic level metrics **/
+    public MetricName topicRecordSendRate(Map<String, String> metricTags) {
```
Nit: should this be simply `tags` since we call them `tags` in createMetricName. Using a different name can make it sound like there's a difference.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/SenderMetricsRegistry.java:
```diff
@@ -16,134 +16,170 @@
+        this.topicTags = new HashSet<String>(tags);
+        this.topicTags.add("topic");
+
+        this.topicRecordSendRate = createTemplate("record-send-rate", TOPIC_METRIC_GROUP_NAME, "The average number of records sent per second for a topic.", topicTags);
```
Worth adding a comment explaining why we can't create the `metricName` for these (i.e. we don't know the topic name).

ijuma reviewed (commented):
Thanks for the PR, left a few comments.

wushujames committed:
These variables never change, so declare them as final.

wushujames committed:
Inline these metrics.

wushujames committed:
Rename some variables and methods, per PR feedback.

wushujames committed:
In order to reduce boilerplate code, use a helper method for the client-level metrics, and another helper method for the per-topic metrics

wushujames committed:
Wrap lines, per PR feedback.

wushujames committed:
Merge remote-tracking branch 'origin/trunk' into producer_sender_metrics_docs_different

wushujames commented:
@ijuma: Thanks for the feedback. I updated the PR with all your suggestions.

ijuma reviewed (approved):
Thanks, LGTM.
